applications/nrf9160dk_demo/src/main.c

```diff
@@ -0,0 +1,1029 @@
+	k_work_init(&send_msg_data_work, send_msg_data_work_fn);
+	k_work_init(&send_modem_at_cmd_work, send_modem_at_cmd_work_fn);
+	k_delayed_work_init(&cloud_reboot_work, cloud_reboot_handler);
+	k_delayed_work_init(&cycle_cloud_connection_work,
+			    cycle_cloud_connection);
+	k_work_init(&device_status_work, device_status_send);
+#if CONFIG_MODEM_INFO
+	k_delayed_work_init(&rsrp_work, modem_rsrp_data_send);
+#endif /* CONFIG_MODEM_INFO */
+}
+
+/**@brief Configures modem to provide LTE link. Blocks until link is
+ * successfully established.
+ */
+static void modem_configure(void)
+{
+#if defined(CONFIG_BSD_LIBRARY)
+	if (IS_ENABLED(CONFIG_LTE_AUTO_INIT_AND_CONNECT)) {
+		/* Do nothing, modem is already turned on
+		 * and connected.
+		 */
+	} else {
+		int err;
+
+		LOG_INF("Connecting to LTE network. ");
+		LOG_INF("This may take several minutes.");
+		ui_led_set_pattern(UI_LTE_CONNECTING);
+
+		err = lte_lc_init_and_connect();
+		if (err) {
+			LOG_ERR("LTE link could not be established.");
+			error_handler(ERROR_LTE_LC, err);
+		}
+
+		LOG_INF("Connected to LTE network");
+		ui_led_set_pattern(UI_LTE_CONNECTED);
+	}
+#endif
+}
+
+static void button_sensor_init(void)
+{
+	button_cloud_data.type = CLOUD_CHANNEL_BUTTON;
+	button_cloud_data.tag = 0x1;
+}
+
+#if CONFIG_MODEM_INFO
+/**brief Initialize LTE status containers. */
+static void modem_data_init(void)
+{
+	int err;
+
+	err = modem_info_init();
+	if (err) {
+		LOG_ERR("Modem info could not be established: %d", err);
+		return;
+	}
+
+	modem_info_params_init(&modem_param);
+
+	signal_strength_cloud_data.type = CLOUD_CHANNEL_LTE_LINK_RSRP;
+	signal_strength_cloud_data.tag = 0x1;
+
+	modem_info_rsrp_register(modem_rsrp_handler);
+}
+#endif /* CONFIG_MODEM_INFO */
+
+/**@brief Initializes the sensors that are used by the application. */
+static void sensors_init(void)
+{
+#if CONFIG_MODEM_INFO
+	modem_data_init();
+#endif /* CONFIG_MODEM_INFO */
+
+	k_work_submit_to_queue(&application_work_q, &device_status_work);
+
+	button_sensor_init();
+}
+
+/**@brief User interface event handler. */
+static void ui_evt_handler(struct ui_evt evt)
+{
+	button_send(evt.button, evt.type == UI_EVT_BUTTON_ACTIVE ? 1 : 0);
+}
+
+void handle_bsdlib_init_ret(void)
+{
+	#if defined(CONFIG_BSD_LIBRARY)
+	int ret = bsdlib_get_init_ret();
+
+	/* Handle return values relating to modem firmware update */
+	switch (ret) {
+	case MODEM_DFU_RESULT_OK:
+		LOG_INF("MODEM UPDATE OK. Will run new firmware");
+		sys_reboot(SYS_REBOOT_COLD);
+		break;
+	case MODEM_DFU_RESULT_UUID_ERROR:
+	case MODEM_DFU_RESULT_AUTH_ERROR:
+		LOG_ERR("MODEM UPDATE ERROR %d. Will run old firmware", ret);
+		sys_reboot(SYS_REBOOT_COLD);
+		break;
+	case MODEM_DFU_RESULT_HARDWARE_ERROR:
+	case MODEM_DFU_RESULT_INTERNAL_ERROR:
+		LOG_ERR("MODEM UPDATE FATAL ERROR %d. Modem failiure", ret);
+		sys_reboot(SYS_REBOOT_COLD);
+		break;
+	default:
+		break;
+	}
+	#endif /* CONFIG_BSD_LIBRARY */
+}
+
+void main(void)
```

Comment:
i think it would be helpful to add some brief comments on some of the things done in main.
- why the application work queue is used.
- what bsdlib is doing and how it relates to modem interactions.  would someone ever build/run this without bsdlib enabled?
- what is the watchdog timeout value set to?  is this a kconfig option?

i suppose it could be done in some other doc or readme, but i find it more useful in the source (especially for an example app like this).

Reply:
yes, agreed

Reply:
Finally getting back to this.  I removed the watchdog timeout stuff from this a while ago, to keep it simple.  I'm adding more comments about the work queue etc.

I've incorporated your code from #2177 as well as other improvements in asset tracker that have been merged.  I do like how this has simplified main(), but it also has made this demo's main longer; so while it will be more robust, it will be less simple.